src/utils/format.ts

```diff
@@ -38,6 +38,18 @@ function _decomposeBn(amount: BN, amountPrecision: number, decimals: number): {
   return { integerPart, decimalPart }
 }
 
+export function formatAmount(
```

Comment:
why so many `formatAmount` declarations?

Reply:
That comes from `TokenRow.tsx::L152`:
```ts
setValue(inputId, formatAmountFull(balance.totalExchangeBalance, balance.decimals, false), true)
```
Now that `react-form-hooks` is more strict with its types it only accepts `string` in `setValue(inputId, string here)`
But `formatAmountFull` returns `string | null`. We know from how that function is written that it would return `string` for input of `amount: BN` and `null` for `amount: null | undefined`, we just need to write types of shape `BN => string; null | undefined => null`

That's where function overloading comes in. We create two distinct declarations:
```ts
function formatAmountFull(amount: BN, amountPrecision?: number, thousandSeparator?: boolean): string
// which allows for
formatFullAmount(amountBN): string

//and
function formatAmountFull(amount?: undefined): null
/// which allows for
fotmatFullAmount(): null
formatFullAmount(undefined): null
```
Same for `formatAmount`

And now in case of
```ts
setValue(inputId, formatAmountFull(balance.totalExchangeBalance, balance.decimals, false), true)
```
as long as (and there it always is) `balance.totalExchangeBalance: BN (not undefined)` we always get a `string` in return (previously was `string | null`)

https://www.typescriptlang.org/docs/handbook/functions.html#overloads

Don't you love Typescript? :smile:

Reply:
eheh, yes. It looks like this is "whitebox type definition", where you add the semantics because you know the actual implementation.

Thanks a lot for the detailed explanation!